### PR TITLE
TST: Add test inconsistency in group by (#44803)

### LINKED ISSUE #44803: BUG: `groupby.apply()` inconsistently indexes its return value with the groupby-cols
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame(dict(
    name=["Alice", "Bob", "Carl", "Dan", "Eve"],
    age=[20, 21, 20, 21, 20],
    height=[165, 175, 189, 182, 157],
    weight=[65, 72, 95, 83, 58],
)).set_index("name")

indexed_by_age = df.groupby(["age"]).apply(lambda group: group.copy())
print(indexed_by_age)

print()
print("=" * 30)
print()

indexed_only_by_name = df.groupby(["age"]).apply(lambda group: group)
print(indexed_only_by_name)
```

### Issue Description

The above snippet produces the following output:

```
           age  height  weight
age name
20  Alice   20     165      65
    Carl    20     189      95
    Eve     20     157      58
21  Bob     21     175      72
    Dan     21     182      83

==============================

       age  height  weight
name
Alice   20     165      65
Bob     21     175      72
Carl    20     189      95
Dan     21     182      83
Eve     20     157      58
```

Where we can see that despite the return values of the `apply-func` being semantically identical (`group` vs. `group.copy()`), one ends up being indexed with `age` and `name` while the other only gets `name`.

### Expected Behavior

I would have expected the two returned dataframes to be the same. I have a personal preference for the first option, with the `age` appearing in the index, but I think I would be able to work with both.

-----

I know other inconsistencies have been reported:
- #22545
- #22546

But IMHO never as clearly as this. I'm hoping this reproducer helps diagnose the issue more easily.
Interestingly, the difference only creeps up when `grouping_keys` is set to `True`. Otherwise, we get the second output twice.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.9.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.15.6-arch2-1
Version          : #1 SMP PREEMPT Thu, 02 Dec 2021 15:47:09 +0000
machine          : x86_64
processor        :
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.21.4
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 58.5.3
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by NumberPiOso**
- [X] closes #44803
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -32,6 +32,20 @@
...
+    df = DataFrame(
+        {
+            "name": ["Alice", "Bob", "Carl"],
+            "age": [20, 21, 20],
+        }
+    ).set_index("name")
+
+    grp_by_same_value = df.groupby(["age"]).apply(lambda group: group)
+    grp_by_copy = df.groupby(["age"]).apply(lambda group: group.copy())
+    assert grp_by_same_value == grp_by_copy
```
Comment (mroeschke) [inline pandas/tests/groupby/test_groupby.py:46]: Please use `tm.assert_frame_equal`

Comment (NumberPiOso) [reply pandas/tests/groupby/test_groupby.py]: Thank you for the corrections :thumbsup:

**Comment by NumberPiOso** [comment]
The code fails one test, but the test it's named _pytest (actions-310-numpydev.yaml, not slow and not network, xclip, deprecate, -W error)_.

Can someone help me with the error?

**Comment by mroeschke** [comment]
The failure is unrelated. Could you merge main again?

**Comment by NumberPiOso** [comment]
Done !

**Comment by jreback** [comment]
thanks @NumberPiOso

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -32,6 +32,20 @@
 from pandas.core.groupby.base import maybe_normalize_deprecated_kernels
 
 
+def test_group_by_copy():
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py:35]: hmm actually should be in `pandas/tests/groupby/test_apply_mutate.py`

@NumberPiOso if you wouldn't mind doing a PR to move this.

Comment (NumberPiOso) [reply pandas/tests/groupby/test_groupby.py:35]: Yes! ,  PR#45578

**Pull request merged**